src/import_edf.jl

```diff
@@ -208,81 +302,264 @@ function groupby(f, list)
+                                   )
+                    return Plan(row)
+                end
+            end
+        end
+    catch e
+        return Plan(_errored_row(row, e))
+    end
 
-`errors` contains `SamplesInfoError`s thrown if channels corresponding to a signal
-were extracted but an error occured while interpreting physical units,
-while promoting sample encodings, or otherwise constructing a `SamplesInfo`.
+    # nothing matched, return the original signal header (as a namedtuple)
+    return Plan(row)
+end
 
-`signal_names` should be an iterable of `String`s naming the signal types to
-extract (e.g., `["ecg", "ekg"]`; `["eeg"]`).
+# create a table with a plan for converting this EDF file to onda: one row per
+# signal, with the Onda.SamplesInfo fields that will be generated (modulo
+# `promote_encoding`).  The column `onda_signal_idx` gives the planned grouping
+# of EDF signals into Onda Samples.
+#
+# pass this plan to edf_to_onda_samples to actually run it
 
-`channel_names` should be an iterable of channel specifications, each of which
-can be either a `String` giving the generated channel name, or a `Pair` mapping
-a canonical name to a list of alternatives that it should be substituted for
-(e.g., `"canonical_name" => ["alt1", "alt2", ...]`).
+"""
+    plan_edf_to_onda_samples(edf::EDF.File;
+                             labels=STANDARD_LABELS,
+                             units=STANDARD_UNITS,
+                             preprocess_labels=(l,t) -> l,
+                             onda_signal_groups=grouper((:kind, :sample_unit, :sample_rate)))
+
+Formulate a plan for converting an `EDF.File` to Onda Samples.  This applies
+`plan_edf_to_onda_samples` to each individual signal contained in the file,
+storing `edf_signal_idx` as an additional column.  The resulting rows are then
+grouped according to `onda_signal_grouper` (by default, the `:kind`,
+`:sample_unit`, and `:sample_rate` columns), and the group index is added as an
+additional column in `onda_signal_idx`.
+
+The resulting plan is returned as a table.  No signal data is actually read from
+the EDF file; to execute this plan and generate `Onda.Samples`, use
+[`edf_to_onda_samples`](@ref)
+"""
+function plan_edf_to_onda_samples(edf::EDF.File;
+                                  labels=STANDARD_LABELS,
+                                  units=STANDARD_UNITS,
+                                  preprocess_labels=(l,t) -> l,
+                                  onda_signal_groups=grouper((:kind, :sample_unit, :sample_rate)))
+    # remove non-Signals (e.g., AnnotationsSignals), keeping track of indices
+    enum_signals = [(i, s) for (i, s) in enumerate(edf.signals) if s isa EDF.Signal]
+    plan_rows = map(enum_signals) do (edf_signal_idx, signal)
+        row = plan_edf_to_onda_samples(signal.header, edf.header.seconds_per_record;
+                   labels, units, preprocess_labels)
+        return rowmerge(row; edf_signal_idx)
+    end
```

Comment:
This was a bit difficult to follow at first. I almost feel like it'd be more straightforward to just to it all in one go:
```suggestion
    plan_rows = [rowmerge(plan_edf_to_onda_samples(s.header, edf.header.seconds_per_record;
                                                   labels, units, preprocess_labels);
                          edf_signal_idx=i)
                 for (i, s) in enumerate(edf.signals)
                 if s isa EDF.Signal]
```